custom-gas-token: add e2e test

geoknee opened the pull request:
Adds a single e2e test asserting on behaviour for both CGT and non-CGT chains. 

Includes utility to activate CGT on a chain which didn't activate the feature at genesis.

Closes https://github.com/ethereum-optimism/client-pod/issues/766

geoknee committed:
add tests for activating custom gas token and bridging tokens to l2

geoknee committed:
update test following concurrent change in SystemConfig ABI

geoknee committed:
update systemconfig bindings

steps to reproduce:

git checkout  v1.7.3 -- op-bindings

remove WETH9 from artifacts.json

cd op-bindings

make bindgen-generate-local

select this one file discard the others

geoknee committed:
use require.NoError instead of require.Nil

geoknee committed:
add assertions for WETH name and symbol getters

tynes committed:
build: fix

geoknee committed:
refactor using setup + subtests

geoknee committed:
minor changes in response to review

geoknee committed:
devnet-allocs: use sender with known private key

some e2e tests require the ability to transact as the system owner safe owner, so we can't use the foundry default sender (private key unknown)

This uses the same address as "Deployer"

geoknee committed:
use deployer as Safe owner

geoknee committed:
WIP refactor test

geoknee committed:
add most of the remaining test assertions

geoknee committed:
rename file

geoknee committed:
flatten out subtests

we want to specify a strict ordering of operations

geoknee committed:
complete checkWithdrawal step

geoknee committed:
reorder test steps

geoknee committed:
check both ETH and token balances on l1 during withdraw

geoknee committed:
simplify callViaSafe helper

geoknee committed:
factor out expectations

geoknee committed:
add some extra boilerplate to track TransactionDeposited event

and GasPayingTokenSet event on L2

geoknee committed:
remove comments

geoknee committed:
remove wayward line

geoknee commented on op-e2e/custom_gas_token_test.go:
```diff
@@ -0,0 +1,468 @@
+		}
+	}
+
+	// Begin by testing behaviour when CGT feature is not enabled
+	enabled := false
+	checkDeposit(t, enabled)
+	checkL1TokenNameAndSymbol(t, enabled)
+	checkL2TokenNameAndSymbol(t, enabled)
+	checkWETHTokenNameAndSymbol(t, enabled)
+
+	// Activate custom gas token feature (devnet does not have this activated at genesis)
+	setCustomGasToken(t, cfg, sys, weth9Address)
+
+	// Now test behaviour given CGT feature is enabled
+	enabled = true
+	checkDeposit(t, enabled)
+	checkWithdrawal(t)
+	checkL1TokenNameAndSymbol(t, enabled)
+	checkL2TokenNameAndSymbol(t, enabled)
+	checkWETHTokenNameAndSymbol(t, enabled)
```
This is pithy / handy summary of what this test does.

geoknee commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -574,7 +574,7 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
                 uint256(0), // mint
                 uint256(0), // value
-                uint64(SYSTEM_DEPOSIT_GAS_LIMIT), // gasLimit
+                uint64(10 * SYSTEM_DEPOSIT_GAS_LIMIT), // gasLimit
```
I believe there is a bug here, the deposit transaction runs out of gas and this causes the test assertions around L1Block and WETH contracts to fail. This may not be the best fix, but it is sufficient to get the tests passing (locally, at least).

tynes replied:
If you could log the gas used of this tx that would be super helpful. It emits an event on L2 that you can search for in the `L1Block` predeploy

geoknee replied:
```
    custom_gas_token_test.go:455: RECEIPT &{Type:126 PostState:[] Status:1 CumulativeGasUsed:146661 Bloom:[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 4 0 0 0 0 0 0 0 0 0 0 0 10 0 0 0 16 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] Logs:[0x14007f92a50] TxHash:0x11eff4337d692df26fa8db89a0764cead90d1adab705afd5cf99debd434dc5e4 ContractAddress:0x0000000000000000000000000000000000000000
    GasUsed:96185 
EffectiveGasPrice:+0 BlobGasUsed:0 BlobGasPrice:<nil> DepositNonce:0x1400862c788 DepositReceiptVersion:0x1400862c790 BlockHash:0x53f3d61d69f57c53e1b0a67511fde32e970bcf68ec3cb3b989435c3ce0f71782 BlockNumber:+22 TransactionIndex:1 L1GasPrice:<nil> L1BlobBaseFee:<nil> L1GasUsed:<nil> L1Fee:<nil> FeeScalar:<nil> L1BaseFeeScalar:<nil> L1BlobBaseFeeScalar:<nil>}
```

geoknee replied:
Should be fixed by https://github.com/ethereum-optimism/optimism/pull/10449

geoknee commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -143,7 +143,7 @@ def devnet_l1_allocs(paths):
     fqn = 'scripts/Deploy.s.sol:Deploy'
     run_command([
-        'forge', 'script', fqn, "--sig", "runWithStateDump()"
+        'forge', 'script', fqn, "--sig", "runWithStateDump()", "--sender", "0x90F79bf6EB2c4f870365E785982E1f101E93b906"
```
We need to set the sender here to an account we know the private key of, because the sender ends up being the owner of the ProxyAdmin SAFE (which we need to enable the CGT feature).

tynes replied:
Oh i see, apologies for this, could you include a comment here saying that?

geoknee replied:
[94f901a](https://github.com/ethereum-optimism/optimism/pull/10440/commits/94f901a44406e860f867a173f9004054d237d638)

geoknee committed:
flatten out sub test

geoknee committed:
add comment to devnet forge script command

tynes commented:
Copying over from the previous PR just for visibility


Basically we want to ensure there are not any reversions in behavior specifically around the weth predeploy since this code will run on OP Mainnet

For non custom gas token chains:

- `OptimismPortal.depositERC20Transaction()` reverts
- `L1Block.gasPayingToken` returns `(ETHER_TOKEN_ADDRESS, 18)`
- `L1Block.gasPayingTokenSymbol` returns `ETH`
- `L1Block.gasPayingTokenName` returns `Ether`
- `SystemConfig.gasPayingToken` returns `(ETHER_TOKEN_ADDRESS, 18)`
- `SystemConfig.gasPayingTokenSymbol` returns `ETH`
- `SystemConfig.gasPayingTokenName` returns `Ether`
- `WETH` predeploy returns `name()` == `"Wrapped Ether"` and `symbol()` returns `"WETH"`

For custom gas token chains:

- basic lock erc20+mint flow
    - validate native asset balance of `to` address changes on L2 by `value`
- basic withdraw flow for erc20s
    - user withdraws native token (call `L2ToL1MessagePasser` directly)
    - gets custom erc20 token on L1
- `WETH` predeploy returns the correct name + symbol, should be prefixed with `Wrapped` and `W`
- `L1Block.gasPayingToken`, `L1Block.gasPayingTokenName`, `L1Block.gasPayingTokenSymbol` all correct, compared against values in ERC20 token

tynes commented:
If you rebase on https://github.com/ethereum-optimism/optimism/pull/10449 then it should fix the issue you found

geoknee committed:
remove version check on systemConfig

geoknee committed:
update systemconfig bindings

steps to reproduce:

git checkout  v1.7.3 -- op-bindings

remove WETH9 from artifacts.json

cd op-bindings

make bindgen-generate-local

select this one file discard the others

geoknee committed:
skip TestCustomGasToken when UseFPAC is true

geoknee committed:
use differente helper to skip test

geoknee commented on op-e2e/custom_gas_token_test.go:
```diff
@@ -0,0 +1,464 @@
+)
+
+func TestCustomGasToken(t *testing.T) {
+	InitParallel(t, SkipOnFPAC) // Custom Gas Token feature is not yet compatible with FPAC
```
Drawing attention here.

geoknee committed:
harmonize use of context.Background{}

geoknee committed:
tidy up and add comments

tynes commented on op-e2e/custom_gas_token_test.go:
```diff
@@ -0,0 +1,456 @@
+	InitParallel(t, SkipOnFPAC) // Custom Gas Token feature is not yet compatible with FPAC
+
+	cfg := DefaultSystemConfig(t)
+	cfg.DeployConfig.FinalizationPeriodSeconds = 2 // 2s finalization period
```
i don't think modifying the finalization period does anything as its a contracts thing and the contract has already been deployed. the deploy config is a leaky abstraction in practice

geoknee replied:
Gotcha. I had copy pasted this from https://github.com/ethereum-optimism/optimism/blob/ac0929bf84230e485b1e2c4264b2f52fecb9654e/op-e2e/system_test.go#L1061-L1062 . It doesn't seem to affect the duration of the test, so I removed it.

geoknee replied:
Gotcha. I had copy pasted this from https://github.com/ethereum-optimism/optimism/blob/ac0929bf84230e485b1e2c4264b2f52fecb9654e/op-e2e/system_test.go#L1061-L1062 . It doesn't seem to affect the duration of the test, so I removed it here [701aabd](https://github.com/ethereum-optimism/optimism/pull/10440/commits/701aabd333b1d421a05f7e08e871e0ab53a1664b).

tynes commented on op-e2e/custom_gas_token_test.go:
```diff
@@ -0,0 +1,456 @@
+	defer sys.Close()
+
+	log := testlog.Logger(t, log.LevelInfo)
+	log.Info("genesis", "l2", sys.RollupConfig.Genesis.L2, "l1", sys.RollupConfig.Genesis.L1, "l2_time", sys.RollupConfig.Genesis.L2Time)
```
do we need these loglines?

geoknee replied:
I don't think so [fb6a53a](https://github.com/ethereum-optimism/optimism/pull/10440/commits/fb6a53a80f279e8cbb8130ef3151b49594310f4b)

tynes commented on op-e2e/custom_gas_token_test.go:
```diff
@@ -0,0 +1,456 @@
+		common.HexToAddress("0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"),
+		"Ether",
+		"ETH",
+		uint8(0x12),
```
why use hex here?

geoknee replied:
No particular reason, I think when this gets logged out it looks this way. Reverted to decimal here [073b51b](https://github.com/ethereum-optimism/optimism/pull/10440/commits/073b51b0617c1ffd9a4fa6519e228b2818b071c6)

geoknee committed:
remove cfg.DeployConfig.FinalizationPeriodSeconds = 2

geoknee committed:
remove log lines

geoknee committed:
use decimal instead of hex
